### PR TITLE
fix: cast assumed rank arrays to descriptor before comparing

### PR DESCRIPTION
**Pull request by HarshilShah1804**
Towards https://github.com/lfortran/lfortran/issues/8357#issuecomment-3598092653

With this PR, we test assumed rank arrays for probably all kinds of operations. Now, we can find some codes which are using assumed rank arrays and test them to handle the corner cases which, if any, may have been missed now.

### CODE REVIEW

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -11695,6 +11695,22 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
         ASR::expr_t *left = ASRUtils::EXPR(tmp);
         this->visit_expr(*x.m_right);
         ASR::expr_t *right = ASRUtils::EXPR(tmp);
+        if (ASRUtils::is_assumed_rank_array(ASRUtils::expr_type(left))) {
+            std::string array_name = ASRUtils::symbol_name(ASR::down_cast<ASR::Var_t>(left)->m_v);
+            if (assumed_rank_arrays.find(array_name) == assumed_rank_arrays.end()) {
+                diag.add(Diagnostic("Comparison operations are not allowed on assumed-rank arrays ('" + array_name + "')",
+                    Level::Error, Stage::Semantic, {Label("", {x.base.base.loc})}));
+                throw SemanticAbort();
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:11703]: Let's add a test for this error into continue compilation.

File: integration_tests/CMakeLists.txt [hunk 2]
```diff
@@ -761,6 +761,7 @@ RUN(NAME select_rank_04 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc fortran)
 RUN(NAME select_rank_05 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc fortran)
 RUN(NAME select_rank_06 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
 RUN(NAME select_rank_07 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
+RUN(NAME select_rank_08 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc EXTRA_ARGS --realloc-lhs-arrays)
```
Comment (certik) [inline integration_tests/CMakeLists.txt:764]: Why is `--realloc-lhs-arrays` needed? I don't see any allocatable arrays in the code.

Comment (HarshilShah1804) [reply integration_tests/CMakeLists.txt:764]: We create a temporary after the `array_struct_temporary` named `_libasr_integer_compare_left` which is an allocatable array. In other cases we would have dimensions available for the allocation, so we insert an allocate statent ourselves but for assumed rank arrays we didn't had the dimensions so we don't insert the allocate statement. So, I had used the `--realloc-lhs-arrays`. 

But I think your concern is right. I'll check for the implementation on how we handle deferred shape arrays like `a(:)`. Mostly it would be involving using the runtime dimensions using the `size()` function. 

What do you suggest shall we meege this now and correct for the inserting the allocate statement using runtime size later. Or shall we fix this now only and then merge this. Please give your suggestions on how to move forward with this. I believe it is a small fix and not much difficult to handle.

File: src/lfortran/semantics/ast_common_visitor.h (hunk 1 above)
Comment (HarshilShah1804) [reply src/lfortran/semantics/ast_common_visitor.h:11703]: Oh right. I will add it in some time.

File: integration_tests/CMakeLists.txt (hunk 2 above)
Comment (certik) [reply integration_tests/CMakeLists.txt:764]: We can merge this, and fix later. Go ahead and open up issues. Let's also run full tests first.

Comment (HarshilShah1804) [reply integration_tests/CMakeLists.txt:764]: Perfect.

Comment (HarshilShah1804) [reply integration_tests/CMakeLists.txt:764]: I have opened #9165 for that.

File: src/lfortran/semantics/ast_common_visitor.h (hunk 1 above)
Comment (HarshilShah1804) [reply src/lfortran/semantics/ast_common_visitor.h:11703]: Since, half of the CI checks have completed and others are completed about 50 %, is it okay if I just add this in a subsequent PR.

**Comment by HarshilShah1804** [comment]
I have the fix for the `--realloc-lhs-arrays` ready on my local branch. I will push it and open the PR once this gets merged. I will also add the error test in that PR.
